jetty-core/jetty-websocket/jetty-websocket-core-common/src/main/java/org/eclipse/jetty/websocket/core/internal/PerMessageDeflateExtension.java

```diff
@@ -264,57 +263,37 @@ public void demand()
         incomingFlusher.demand();
     }
 
-    private class OutgoingFlusher extends TransformingFlusher
+    private class OutgoingFlusher extends WebSocketFlusher
     {
-        private boolean _first;
-        private Frame _frame;
-
         @Override
-        protected boolean onFrame(Frame frame, Callback callback, boolean batch)
+        protected boolean onFrame(OutgoingEntry entry, boolean first)
         {
-            if (frame.isControlFrame())
+            if (first)
             {
-                OutgoingEntry entry = new OutgoingEntry.Builder(getCurrentEntry())
-                    .frame(frame)
-                    .callback(callback)
-                    .batch(batch)
-                    .build();
-                nextOutgoingFrame(entry);
-                return true;
-            }
-
-            _first = true;
-            _frame = frame;
-
-            // Provide the frames payload as input to the Deflater.
-            getDeflater().setInput(frame.getPayload().slice());
-            callback.succeeded();
-            return false;
-        }
+                if (entry.getFrame().isControlFrame())
+                {
+                    nextOutgoingFrame(entry);
+                    return true;
+                }
 
-        @Override
-        protected boolean transform(Callback callback)
-        {
-            boolean finished = deflate(callback);
-            _first = false;
+                // Provide the frames payload as input to the Deflater.
+                getDeflater().setInput(entry.getFrame().getPayload().slice());
```

Comment:
Do you need slicing here?

Reply:
We do because currently the contract is that we do not consume the frame payload on a sendFrame (see https://github.com/jetty/jetty.project/issues/13290).